lib/bundler/lockfile_parser.rb

```diff
@@ -101,17 +101,16 @@ def warn_for_outdated_bundler_version
       return unless bundler_version
       prerelease_text = bundler_version.prerelease? ? " --pre" : ""
       current_version = Gem::Version.create(Bundler::VERSION)
-      case current_version.segments.first <=> bundler_version.segments.first
-      when -1
-        raise LockfileError, "You must use Bundler #{bundler_version.segments.first} or greater with this lockfile."
-      when 0
-        if current_version < bundler_version
-          Bundler.ui.warn "Warning: the running version of Bundler (#{current_version}) is older " \
-               "than the version that created the lockfile (#{bundler_version}). We suggest you to " \
-               "upgrade to the version that created the lockfile by running `gem install " \
-               "bundler:#{bundler_version}#{prerelease_text}`.\n"
-        end
+      current_major_version = current_version.segments.first
+      major_bundler_version = bundler_version.segments.first
+      if current_major_version < major_bundler_version && current_major_version > 2
```

Comment:
Should be `major_bundler_version > 2`, right?

Reply:
I had the exact same thought, but maybe it's because of the naming confusion? I had a second look before and decided it was correct.

Reply:
`current_major_version` is the Bundler (major) version, right?
And `major_bundler_version` is the (major) `BUNDLED WITH` version.
We want to error out if Bundler version < `BUNDLED WITH` and `BUNDLED WITH` > 2 (all `BUNDLED WITH` <= 2 lockfiles are compatible).

I believe if Bundler version < `BUNDLED WITH` and Bundler version > 2 is incorrect, for instance for Bundler version 2 (or 1) and `BUNDLED WITH` 3, which should error out but this condition is `false`.

Reply:
Right, this is so confusing... The thing (I believe) is that the lower left part of the table should be already caught by the rubygems. The error would be:

> Could not find 'bundler' (3) required by your Gemfile.lock. To update to the latest version installed on your system, run `bundle update --bundler`. To install the missing version, run `gem install bundler:3`

Downgrading that to a warning as well is the complementary part of this in rubygems: https://github.com/rubygems/rubygems/pull/2696.

This brings up other issues:

* We should probably remove the `bundle update --bundler` recommendation from that advice, since I don't think `bundle update --bundler` works for downgrading the `BUNDLED WITH` version.

* We might want to remove the check in those cases, because otherwise, with both PRs applied, you get duplicated warnings, like 

```
$ bundle 
Could not find 'bundler' (3.0.0) required by your /home/deivid/Code/playground/bundler/bundle_with_dance/Gemfile.lock.
To update to the latest version installed on your system, run `bundle update --bundler`.
To install the missing version, run `gem install bundler:3.0.0`

Warning: the running version of Bundler (2.1.0.pre.1) is older than the version that created the lockfile (3.0.0).
We suggest you to upgrade to the version that created the lockfile by running `gem install bundler:3.0.0`.
The Gemfile specifies no dependencies
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
```

Reply:
Maybe we should actually remove this check altoghether from `bundler` and unify everything in `rubygems`.